python/kwiver/sprokit/adapters/adapter_data_set.cxx

```diff
@@ -0,0 +1,131 @@
+/*ckwg +29
+ * Copyright 2020 by Kitware, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  * Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  * Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <sprokit/processes/adapters/adapter_data_set.h>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl_bind.h>
+#include <memory>
+
+// Type conversions
+#include <vital/types/image_container.h>
+#include <vital/types/detected_object_set.h>
+#include <vital/types/descriptor_set.h>
+#include <vital/types/track_set.h>
+#include <vital/types/object_track_set.h>
+#include <vital/types/feature_track_set.h>
+#include <vital/types/timestamp.h>
+#include <vital/types/geo_polygon.h>
+#include <vital/types/homography_f2f.h>
+
+using namespace pybind11;
+using namespace kwiver::adapter;
+
+PYBIND11_MODULE(adapter_data_set, m)
+{
+  enum_<adapter_data_set::data_set_type>(m, "DataSetType"
+    , "Type of data set.")
+    .value("data", adapter_data_set::data)
+    .value("end_of_input", adapter_data_set::end_of_input)
+  ;
+
+  class_< adapter_data_set, std::shared_ptr<adapter_data_set > >(m, "AdapterDataSet")
+    .def_static("create", &adapter_data_set::create
```

Comment:
Does it make sense to add a `__str__` and `__repr__`  to this class? Not sure what the purpose of an AdapterDataSet is. It might also be nice to have some a class-level docstring with an "Example" doctest block if it is something that we expect users to work with. Note that `xdoctest` supports executing doctests within pybind11 modules, so the example will help with CI testing. I have another class that I did this on if you need an example (see `vital/types/image.cxx`).

Reply:
I'm not so sure about `__str__` and `__repr__`. I think they would be good to eventually have, since adapter_data_set (ADS) is essentially a map/python dictionary, but implementing it right now would be difficult and not very beneficial. We would essentially have to loop through the ADS ports/values, call `get_datum<vital::any>` on the values to avoid having to deal with specific template types, cast those to py::objects, then call py::str() on the casted values (if that will work). A fair amount of vital types don't have `__str__` or operator<< (this is still a WIP), so we would likely be getting things along the lines of "timestamp object at 0x....." most of the time.

Reply:
Docstrings added in 4c92416d1d4373d10a7e85ded265c3b5aef134fe

Reply:
We should maintain an issue where we outline "Pythonic" improvements that could be made to these bindings. 

If ADS is effectively a dict, then it should have a dict-like interface with    `__iter__`, `__len__`, `__contains__`, `__getitem__`, `items`, `keys`, `values`, `get`,  and if it should be writable it should have:   `__delitem__`, `__setitem__`, `update`, and perhaps: `copy`. 

For `__str__` and `__repr__` I think its useful for objects like this to behave as a dict  in that respect in order to make debug outputs readable. There are exceptions to this rule though, like if we expect ADS to contain a ton of info, then I typically like to print a summary like number of items. In general though I think its always good practice for a `__str__` and `__repr__` to give some indication as to the contents of an object. The default implementations are very restrictive. 

In this case, I think the keys might have important information even if the values were relatively sparse.

Reply:
I've created a few issues for vital types that could use some extra pythonic functionality. I think this would be a good idea in this case.

As for str and repr, we may be able to print the type of the value for each key.

Reply:
`__iter__` was present in the original bindings. In 1c40fdd8f27f207f3620e362328f94bf75ee9e11, I added `__str__`, `__repr__`, `__nice__`, `__getitem__`, `__setitem__` and `__len__`. Most of the rest of those would require changes to the actual adapter_data_set class to be efficient (simply wrapping std::map's functions). If you'd prefer that I add the rest I would be happy to do so, otherwise I can create an issue for the remaining methods.